[JENKINS-43322] Nested views should be also searchable by name

szpak opened the pull request:
# Description

See [JENKINS-43322](https://issues.jenkins-ci.org/browse/JENKINS-43322).

**Details**: As of Jenkins 2.53 searching for view by name works only for top level views. It can be frustrating when you use nested views especially in companion with Delivery Pipeline Plugin - nested pipelines are not searchable.
Looking in the code the reason is to have the "views" field in Jenkins only references to top level views. Nested ones (ViewGroup) are ignored in getViews() and getView(name). Proposed solution adds an extra method getAllViews() (a similar convention to getItems() and getAllItems()) which returns all views recursively and enhance getView(name) to use it.

~Please note. Only the [last commit](https://github.com/jenkinsci/jenkins/pull/2831/commits/4e9358a2d90419b29d7dfeada88d2937536b880c) contains changes - the others are related to a Jenkins switch to Java 8 which is used to simplify implementation (#2802) -  ~ DO NOT MERGE~. - rebased against master - just one commit

### TODO:
 - [x] Wait for #2802 with Java 8 switch to be merged (planed for that week)
 - [x] Better test coverage (at which level? how would be best to created nested views in test?)

### Open issues/questions:
 - [ ] Should `getAllViews()` be exported?
 - [x] ~Replace `ViewGroupMixIn` with default methods in the interface~ - another PR

### Changelog entries

Proposed changelog entries:

* [JENKINS-43322](https://issues.jenkins-ci.org/browse/JENKINS-43322) Nested views are now searchable by name

### Submitter checklist

- [x] JIRA issue is well described
- [x] Link to JIRA ticket in description, if appropriate
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc)

### Desired reviewers

TODO

Commit:
[JENKINS-43322] Nested views searchable by name

szpak commented:
CC @jenkinsci/code-reviewers as #2802 has been merged.

szpak commented:
Added test cases. It is ready for review.

daniel-beck commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -1811,6 +1811,11 @@ public View getView(String name) {
         return viewGroupMixIn.getViews();
     }
 
+    @Exported     //Just to be exported - @Exported on default methods in interface is not supported
```
Why export? Perhaps with a lower visibility?

jvz replied:
I think you're right, @szpak

batmat replied:
Dummy question, is adding this required for the whole improvement to work? If not, I'd strongly recommend to file it in a separate PR to unblock the PR and have a dedicated discussion. (I suspect it is needed, but then at least this will be clarified, hopefully not just for me)

daniel-beck replied:
This is not required to make this work, so should be removed. This looks like a separate and mostly unrelated change.

batmat replied:
removed @daniel-beck

szpak commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -1811,6 +1811,11 @@ public View getView(String name) {
         return viewGroupMixIn.getViews();
     }
 
+    @Exported     //Just to be exported - @Exported on default methods in interface is not supported
```
It's the open question. Probably it would be best to have `allViews` exported instead of just the 1st level views the same as (effectively) all jobs are exported (being top level jobs). Nevertheless for backward compatibility `views` probably should be exported anyway.

@daniel-beck What visibility should be used to have it accessible via remote API, but on the other hand suppressed when root Jenkins elements are listed (unfortunately it's the first level element)?

oleg-nenashev replied:
👎  for that in the current state. The method should conduct `View.READ` permission checks at least.

szpak replied:
@oleg-nenashev I was thinking about that, but the current implementation does permission checks implicitly as it [uses](https://github.com/jenkinsci/jenkins/pull/2831/files#diff-9767005f02cfa46f2643177f75ec4d8bR70) `getViews()` under the hood to get views which should already returns filtered results.

Do you think there should be a double check on received views?

szpak committed:
[JENKINS-43322] Use unfiltered list in getView(name)

getAllViews() returns views filtered by permissions. It is problematic
for example when trying to add a new view with conflicting name.

oleg-nenashev commented on core/src/main/java/hudson/model/ViewGroup.java:
```diff
@@ -58,6 +59,26 @@
+     * @return
+     *      can be empty but never null.
+     *
+     * @since 2.55
```
Please do not explicitly specify versions next time. `@since TODO` is preferable

szpak replied:
I planned to amend it if merged after 2.55. I can change it to TODO.

oleg-nenashev commented on core/src/main/java/hudson/model/TreeView.java:
```diff
@@ -140,9 +140,16 @@ public void deleteView(View view) throws IOException {
+        for (View v : views) {
+            if (v.getViewName().equals(name))
                 return v;
+            if (v instanceof ViewGroup) {
```
The only problem with this approach is that Nested views may appear as a search result even if there is TopLevel view with such name. And as a user I would expect them to have higher priority. Maybe it worth checking all TopLevel views first.

I would also vote for making a parameterizable method `View getView(String name, boolean recursive)` and maybe changing only the API calls which are really related to search. Otherwise it it hard to predict the impact of the defaukt behavior change

batmat replied:
> And as a user I would expect them to have higher priority.

FWIW, I checked that the selection takes you to the top level view, so I think the current behavior is fine.

oleg-nenashev commented on core/src/main/java/hudson/model/ViewGroupMixIn.java:
```diff
@@ -93,8 +89,16 @@ public synchronized void deleteView(View view) throws IOException {
                 return v;
+            }
+            //getAllViews() cannot be used as it filters jobs by permission which is bad e.g. when trying to add a new job
+            if (v instanceof ViewGroup) {
```
same as above

oleg-nenashev reviewed (dismissed):
I am against merging REST API method in the current state.

I would also propose to implement API in a different way to avoid collateral damage and performance degradation in `getView` users. But it just needs a discussion

szpak commented on core/src/main/java/hudson/model/TreeView.java:
```diff
@@ -140,9 +140,16 @@ public void deleteView(View view) throws IOException {
+        for (View v : views) {
+            if (v.getViewName().equals(name))
                 return v;
+            if (v instanceof ViewGroup) {
```
I originally implemented it as an extra loop, but later inlined it. However, I forgot about the order in the search results which sounds very convincing for me. I will restore the old implementation.

szpak replied:
Regarding `recursive` I can add it, however, in most cases recursive search should be a desired behavior. Nevertheless, I don't know Jenkins internals and user use cases, so I'm not against that change and can implement it.

szpak committed:
[JENKINS-43322] Top level views returned first

When search by name. As suggested in the review.

szpak commented:
Thanks for your review @oleg-nenashev . I implemented your suggestions except for `recursive`, but I'm getting more and more convinced that is it a good idea. It seems that currently Jenkins allows to create multiple nested views with the same name. Probably it wasn't done intentionally, but that change can break some instances. I will analyze what other places need to be changed.

oleg-nenashev commented:
@szpak Please let me know once an additional review is required

szpak commented:
@oleg-nenashev I will probably handle `recursive` on Monday. In the meantime what do you think about exporting `allViews`? Initially I wanted to have it available on demand, but it is a first level element and I'm afraid `@Export` cannot be configured to hide it when `jenkins` root element is fetched via remote API.

szpak commented:
Unfortunately a variant with an extra recursive method is problematic in implementation.
1. The new method cannot be simply added to `ViewGroup` to do not break plugins with old implementation.
2. Default implementation of `getView(name, recursive)` (in the interface) could in non recursive mode delegate to old `getView(name)` and fail in runtime with information that given class does not support recursive search. However, as we cannot use `getAllViews()` for recursive search (as it filters views based on permission) it is needed to ask all nested `ViewGroup` implementations with `getView(name, recursive)`. Effectively having any implementation without an explicit recursive search support would break the whole search mechanism (a runtime exception `UnsupportedOperationException` would be thrown) or those nested views would be silently ignored (which doesn't make sense as current version of for example `NestedView` doesn't support that feature and new versions would need to depend on the current Jenkins version).

As a result, I don't have a vision how to implement a separate recursive search without breaking backward compatibility. @oleg-nenashev maybe you have some better ideas.

jvz commented on core/src/main/java/hudson/model/ViewGroup.java:
```diff
@@ -57,6 +58,26 @@
+     *
+     * @since TODO
+     */
+    default Collection<View> getAllViews() {
```
```suggestion
    default @Nonnull Collection<View> getAllViews() {
```

daniel-beck replied:
This should be addressed.

batmat replied:
done

batmat committed:
Merge branch 'master' into PR-2831-szpak

# Conflicts:
#	core/src/main/java/hudson/model/ViewGroupMixIn.java
#	test/src/test/java/hudson/model/ViewTest.java

batmat commented:
I resolved the conflict to help @szpak, but I think I missed a step since seems it does not even compile... Sorry, looking :man_facepalming:

batmat commented:
@oleg-nenashev I'm not sure your concerns have been addressed or not. Can you please review the current state to see if you are still "requesting change" here? Thanks

batmat reviewed (approved):
Tested the change locally, and it looks good to me.
I also tried the behavior with a nested view with the same name (even with spaces) as the parent one: it shows only one entry (which might be slightly misleading, but that's fine I think), and if you select it this goes to the top one. Which sounds like the least surprising behavior so :+1: 

(For anyone willing to play with this:
`docker run -ti -e ID=2831 -e MERGE_WITH=master -p 8080:8080 jenkins/core-pr-tester`
Then finish the setup and install the `Nested View` Jenkins plugin)

szpak commented:
Thanks @batmat for bringing it back to life :).

As in the meantime (2 years :) ) I stopped working with that Jenkins instance with 8K+ jobs generated with Job DSL where the nested views was really important to have to something much smaller (and with Jenkins Pipeline) I'm not longer up-to-date with the changes I made then (and possible related issues). However, I'm really glad that it still works after your recent changes and definitely it would be good to see it available in the vanilla package :).

batmat commented:
@oleg-nenashev you requested a change here. Can you please have a look to confirm your concerns are addressed. This seems they was reading Matt's answer in https://github.com/jenkinsci/jenkins/pull/2831#discussion_r272348883

@daniel-beck do you have an opinion on this PR? (seeing you commented above).

Thanks

batmat committed:
Remove unnecessary new @Export

batmat committed:
Document that return must not be null

daniel-beck commented:
@batmat Without `@Exported` you don't need the override at all.

batmat committed:
:facepalm: now it's just noop code that should be removed

batmat commented:
:man_facepalming: done

batmat commented:
@daniel-beck does it look better now? I think it could be considered ready now isn't it?

batmat commented:
I will likely merge this before Monday to land it in next weekly, if nobody objects in the meantime.
